File: apps/app-frontend/src/helpers/hosting.ts
import { invoke } from '@tauri-apps/api/core'

/**
 * Which server jar a hosted server runs. Only Paper can be given players' real
 * Mojang identities; `vanilla` servers predate that and see offline UUIDs.
 */
export type ServerFlavor = 'vanilla' | 'paper'

export interface HostedServer {
	id: string
	name: string
	directory: string
	mc_version: string
	java_path: string | null
	port: number
	endpoint_name: string | null
	flavor: ServerFlavor
	created: number
	modified: number
}

/** Matches the public suffix every Minekube Connect endpoint resolves under. */
const CONNECT_PUBLIC_SUFFIX = 'play.minekube.net'

export function serverAddress(server: HostedServer): string | null {
	return server.endpoint_name ? `${server.endpoint_name}.${CONNECT_PUBLIC_SUFFIX}` : null
}

export async function listServers(): Promise<HostedServer[]> {
	return await invoke('plugin:hosting|hosting_list_servers')
}

export async function createServer(name: string, version?: string | null): Promise<HostedServer> {
	return await invoke('plugin:hosting|hosting_create_server', { name, version: version ?? null })
}

export async function removeServer(id: string): Promise<void> {
	return await invoke('plugin:hosting|hosting_remove_server', { id })
}

export async function startServer(id: string): Promise<void> {
	return await invoke('plugin:hosting|hosting_start_server', { id })
}

export async function stopServer(id: string): Promise<void> {
	return await invoke('plugin:hosting|hosting_stop_server', { id })
}

export async function serverStatus(id: string): Promise<boolean> {
	return await invoke('plugin:hosting|hosting_server_status', { id })
}

export async function runningServers(): Promise<string[]> {
	return await invoke('plugin:hosting|hosting_running_servers')
}

export async function getLogs(id: string): Promise<string[]> {
	return await invoke('plugin:hosting|hosting_get_logs', { id })
}

export async function sendCommand(id: string, command: string): Promise<void> {
	return await invoke('plugin:hosting|hosting_send_command', { id, command })
}

export async function ensureTunnel(id: string): Promise<string> {
	return await invoke('plugin:hosting|hosting_ensure_tunnel', { id })
}

File: apps/app/src/api/hosting.rs
use crate::api::Result;
use tauri::plugin::TauriPlugin;
use theseus::prelude::*;

pub fn init<R: tauri::Runtime>() -> TauriPlugin<R> {
    tauri::plugin::Builder::new("hosting")
        .invoke_handler(tauri::generate_handler![
            hosting_list_servers,
            hosting_create_server,
            hosting_remove_server,
            hosting_start_server,
            hosting_stop_server,
            hosting_server_status,
            hosting_running_servers,
            hosting_get_logs,
            hosting_send_command,
            hosting_ensure_tunnel,
            hosting_update_server,
            hosting_change_version,
            hosting_set_server_icon,
            hosting_clear_server_icon,
            hosting_get_server_icon,
        ])
        .build()
}

#[tauri::command]
pub async fn hosting_list_servers() -> Result<Vec<HostedServer>> {
    Ok(hosting::list_servers().await?)
}

#[tauri::command]
pub async fn hosting_create_server(
    name: String,
    version: Option<String>,
) -> Result<HostedServer> {
    Ok(hosting::create_server(name, version).await?)
}

#[tauri::command]
pub async fn hosting_remove_server(id: String) -> Result<()> {
    hosting::remove_server(id).await?;
    Ok(())
}

#[tauri::command]
pub async fn hosting_start_server(id: String) -> Result<()> {
    hosting::start_server(id).await?;
    Ok(())
}

#[tauri::command]
pub async fn hosting_stop_server(id: String) -> Result<()> {
    hosting::stop_server(id).await?;
    Ok(())
}

#[tauri::command]
pub async fn hosting_server_status(id: String) -> Result<bool> {
    Ok(hosting::server_status(id).await?)
}

#[tauri::command]
pub async fn hosting_running_servers() -> Result<Vec<String>> {
    Ok(hosting::running_servers().await?)
}

#[tauri::command]
pub async fn hosting_get_logs(id: String) -> Result<Vec<String>> {
    Ok(hosting::get_logs(id).await?)
}

#[tauri::command]
pub async fn hosting_send_command(id: String, command: String) -> Result<()> {
    hosting::send_command(id, command).await?;
    Ok(())
}

#[tauri::command]
pub async fn hosting_ensure_tunnel(id: String) -> Result<String> {
    Ok(hosting::ensure_tunnel(&id).await?)
}

#[tauri::command]
pub async fn hosting_update_server(
    id: String,
    name: String,
    max_players: u32,
    view_distance: u32,
) -> Result<HostedServer> {
    Ok(hosting::update_server(id, name, max_players, view_distance).await?)
}

#[tauri::command]
pub async fn hosting_change_version(
    id: String,
    version: String,
) -> Result<HostedServer> {
    Ok(hosting::change_version(id, version).await?)
}

#[tauri::command]
pub async fn hosting_set_server_icon(
    id: String,
    image: Vec<u8>,
) -> Result<()> {
    hosting::set_server_icon(id, image).await?;
    Ok(())
}

#[tauri::command]
pub async fn hosting_clear_server_icon(id: String) -> Result<()> {
    hosting::clear_server_icon(id).await?;
    Ok(())
}

#[tauri::command]
pub async fn hosting_get_server_icon(id: String) -> Result<Option<String>> {
    Ok(hosting::get_server_icon(id).await?)
}
